File: LeetCode/0124.Binary-Tree-Maximum-Path-Sum/Binary-Tree-Maximum-Path-Sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.res = -sys.maxsize - 1
        self.dfs(root)
        return self.res
    
    def dfs(self, node):
        if not node: return 0
        
        left = self.dfs(node.left)
        right = self.dfs(node.right)
        
        p_sum  = node.val
        if left > 0: p_sum += left
        if right > 0: p_sum += right
        self.res = max(self.res, p_sum)
        
        d_sum = node.val
        if max(left, right) > 0: d_sum += max(left, right)
        return d_sum
        
        

File: LeetCode/0124.Binary-Tree-Maximum-Path-Sum/Binary-Tree-Maximum-Path-Sum.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: An integer
     */
    public int res; 
    public int maxPathSum(TreeNode root) {
        // write your code here
        if(root == null )return 0;
        res = Integer.MIN_VALUE;
        helper(root);
        return res;
    }
    public int helper(TreeNode root){
        if(root == null) return 0;
        int left = Math.max(0, helper(root.left));
        int right = Math.max(0, helper(root.right));
        
        res = Math.max(res, left + right + root.val);
        return Math.max(left, right) + root.val;
    }
}
